File: leetcode_ly/java/List/ReverseLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null) return head; 
        ListNode p = null; 
        ListNode q = head; 
        while(head.next != null){ 
            head = head.next; 
            q.next = p; 
            p = q; 
            q = head; 
        } 
        head.next = p; 
        return head; 
    }
}

File: leetcode_ly/c++/List/ReverseLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == NULL || head->next == NULL){
            return head;
        }
        else{
            ListNode* p = head;
            ListNode* q = head->next;
            ListNode* r;
            head->next = NULL;
            while(q->next){
                r = q->next;
                q->next = p;
                p = q;
                q = r;
            }
            q->next = p;
            head = q;
            return head;
        }
    }
};